src/Interpreters/inplaceBlockConversions.cpp

```diff
@@ -22,46 +23,46 @@ namespace DB
 namespace
 {
 
-ASTPtr defaultRequiredExpressions(Block & block, const NamesAndTypesList & required_columns, const ColumnDefaults & column_defaults)
+/// Add all required expressions for missing columns calculation
+void addDefaultRequiredExpressionsRecursively(Block & block, const String & required_column, const ColumnsDescription & columns, ASTPtr default_expr_list_accum, NameSet & added_columns)
 {
-    ASTPtr default_expr_list = std::make_shared<ASTExpressionList>();
+    checkStackSize();
+    if (block.has(required_column) || added_columns.count(required_column))
+        return;
 
-    for (const auto & column : required_columns)
+    auto column_default = columns.getDefault(required_column);
+
+    if (column_default)
     {
-        if (block.has(column.name))
-            continue;
+        /// expressions must be cloned to prevent modification by the ExpressionAnalyzer
+        auto column_default_expr = column_default->expression->clone();
 
-        const auto it = column_defaults.find(column.name);
+        /// Our default may depend on columns with default expr which not present in block
+        /// we have to add them to block too
+        RequiredSourceColumnsVisitor::Data columns_context;
+        RequiredSourceColumnsVisitor(columns_context).visit(column_default_expr);
```

Comment:
I wish we store defaults as DAG with explicit dependences.
Manipulations with AST here looks kind of weird.